dom.bs

```diff
@@ -2652,6 +2652,63 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
  <!-- Technically this is post-insert. -->
 </ol>
 
+<p>To <dfn export for=Node id=concept-node-ensure-pre-move-validity>ensure pre-move validity</dfn>
+of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run these steps:
+
+<ol>
+ <li>
+  <p>If any of the following conditions are true</p>
+
+  <ul>
+   <li><p><var>parent</var> is <a>connected</a> and <var>node</var> is not <a>connected</a>; or</p></li>
+
+   <li><p><var>parent</var> is not <a>connected</a> and <var>node</var> is <a>connected</a>,</p></li>
+  </ul>
+
+  <p>then <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.
+
+  <p class=note>Moving a node around in a DOM <a>tree</a> requires both it and its eventual parent
+  to both be <a>connected</a>, or both be disconnected. Otherwise, state preservation cannot be
+  guaranteed during the move operation.</p>
+ </li>
+
+ <li><p>If <var>parent</var>'s <a for=/>shadow-including root</a> is not the same as
+ <var>node</var>'s <a for=/>shadow-including root</a>, then <a>throw</a> a
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
+
+ <li><p>If <var>parent</var> is not an {{Element}} or {{DocumentFragment}} <a for=/>node</a>, then
+ <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.
```

Comment:
The reason this restriction is okay is because documents cannot have multiple element children and you can only use `insertBefore()` at the root when the current element is removed. So it really relies on the connected checks above. That's very subtle.

However, when looking at non-elements it's quite a bit more restrictive than what it probably should be. I don't see why we'd forbid moving a comment to be a child of a document. So I would suggest we do this differently, even though in practice there's no benefit to moving comments and the like.

Ideally we'd align this with "pre-insertion validity" or share as much as possible.

Reply:
Yeah, I'd expect this to handle basically ParentNode mixin.

Reply:
Hmm, OK. Would you prefer we simply make this condition perfectly match the pre-insertion validity steps? I.e., expanding it to include Document parents too? I'm fine with that. Let me know what you think @annevk.

> Yeah, I'd expect this to handle basically ParentNode mixin.

Hmm, I'm not exactly sure what this means. But does what I recommend above satisfy what you're looking for?

Reply:
Yeah, I think that would be good, although I have not thought through all implications of such a change.

Reply:
OK I've done that in https://github.com/whatwg/dom/pull/1307/commits/90c9a943a859267acd05b37c48120427e497c7e2, and changed the Chromium implementation and added a test in https://crrev.com/c/6081493. I agree this is a pretty minor change, especially since Documents can still only have a single Element child, and it is nice to have consistency with the pre-insertion validity conditions. I'll resolve this now.